sections/advanced/reaching-ui-state.md

```diff
@@ -0,0 +1,222 @@
+
+Cypress allows to stub all network traffic. We can record the network data from an endpoint, and stub that response every time the UI makes a call to an arbitrary server.
+
+Start by copying the network data from devTools to a json file. Place it in `cypress/fixtures` folder. This folder is made for this purpose, and any reference to it will default to the root of the folder.
+
+![devtools > network tab](../../assets/images/ui-state/devtools-network.PNG)
+
+```javascript
+// json has been copied to a file `cypress/fixtures/agents.json`
+
+// this is a shorthand for cy.fixture(). More at https://docs.cypress.io/api/commands/fixture.html#Accessing-Fixture-Data
+cy.server();
+cy.route('some-xhr-call-that-happens-upon-landing', 'fx:agents.json').as('crutcXHR');
+// all calls to the network route will be stubbed by the data in agents.json file
+```
+
+#### What if there are so many network requests happening?
+Where do we get all our mocks for fixtures? We do not want to manually copy and save them. We want to record them as the test runs against a real API
+
+There are at least 2 Cypress plugins you can utilize for this [1](https://github.com/Nanciee/cypress-autorecord) & [2](https://github.com/scottschafer/cypressautomocker).
+
+If these do not work for you, you can easily create record and playback utilities yourself with these 3 functions.
+
+```javascript
+function stubRecorder(pathToJson) {
+  const xhrData = []; // an empty array to hold the data
+  cy.server({ // if recording, save the response data in the array
+    onResponse: (response) => {
+      const url = response.url;
+      const method = response.method;
+      const data = response.response.body;
+      // We push a new entry into the xhrData array
+      xhrData.push({ url, method, data });
+    }
+  });
+  // this is used as a ‘filter’ for the data you want to record.
+  // You can specify the methods and routes that are recorded
+  cy.log('recording!');
+  cy.route({
+    method: 'GET',
+    url: '*',
+  });
+  // if recording, after the test runs, create a fixture file with the recorded data
+  after(function () {
+    cy.writeFile(`./cypress/fixtures/${pathToJson}.json`, xhrData);
+    cy.log(`Wrote ${xhrData.length} XHR responses to local file ${pathToJson}.json`);
+  });
+}
+
+/** Plays recorded fixture with all required network data as json*/
+function playStubbedFixture(stateFixture) {
+  cy.log(`playing fixture from ${stateFixture}`);
+  cy.fixture(stateFixture, { timeout: 15000 }) // the fixture file may be large and take time in CI
+    .then(data => {
+      for (let i = 0, length = data.length; i < length; i++) {
+        cy.route(data[i].method, data[i].url, data[i].data);
+      }
+    }).as(`stateFixture_stub`);
```

Comment:
it could become a simpler
```js
cy.fixture(stateFixture, { timeout: 15000 }) // the fixture file may be large and take time in CI
  .each(({method, url, data}) => {
    cy.route(method, url, data);
  }).as(`stateFixture_stub`);
```
what do you think?

Reply:
yes, much better that way!